storage/innobase/log/log0log.cc

```diff
@@ -3455,6 +3457,9 @@ logs_empty_and_mark_files_at_shutdown(void)
 					" to be flushed");
 				count = 0;
 			}
+			if (report(&start_time, ut_time())) {
+				my_sd_extend_timeout(INNODB_EXTEND_TIMEOUT_INTERVAL);
+			}
```

Comment:
I would prefer a more fine-grained process here, with some progress reporting of how many pages are to be flushed.
But, we can start with this one. Worst case, this will lie to systemd that shutdown is actively doing something useful, while it is hung due to some problem.

Reply:
couldn't see any easy way with buf_all_free. Is something earlier measuring LSN progress sufficient?

Reply:
Actually, `buf_all_freed()` will always return `TRUE` if it returns. It will crash if any page was not flushed. The code around it could be cleaned up. Yes, the buffer pool scan in `buf_all_freed()` will take some time, but it does not involve any I/O. So, the `if` block immediately above your added patch is dead code. We should just call `buf_all_freed()`; it is essentially an assertion.

We should actually remove this call to `buf_all_freed()`, because a few lines later we would be calling `buf_all_freed()` again. (And by the way, should this call take too much time, systemd is welcome to kill the server, because the redo log checkpoint was already completed.)

The time-consuming part is the call to `log_make_checkpoint_at()`. It waits for the page flushing. The following calls are likely redundant and could be removed:
```c
		fil_flush_file_spaces(FIL_TYPE_TABLESPACE);
		fil_flush_file_spaces(FIL_TYPE_LOG);
```
It looks like we should somehow add timeout extension and progress reporting for `UT_LIST_GET_LEN(buf_pool->flush_list)` (the count should reach zero). The proper point for that would seem to be the function `buf_flush_remove()`.
I would suggest something like this:
```c
if (UNIV_UNLIKELY(srv_shutdown_state == SRV_SHUTDOWN_FLUSH_PHASE)
    && report(…)) {
	my_sd_extend_timeout(INNODB_EXTEND_TIMEOUT_INTERVAL);
```
I think that it might be useful to have a global time base for the shutdown progress reporting and extending the timeout intervals.

Reply:
ok. thanks. Is srv_inc_activity_count usable?

Reply:
Short answer: no.

Long answer: The `srv_inc_activity_count()` seems to affect `srv_check_activity()`. It is used in the 10.1 `buf_flush_page_cleaner_thread()` and the 10.2 `buf_flush_page_cleaner_coordinator()`, but only before shutdown is initiated. Likewise, it is used before shutdown by the purge coordinator thread and the master thread to determine if some more work should be done.

During shutdown the activity count does not seem to play any role.

Reply:
There is a third `buf_all_freed` at the end of the `logs_empty_and_mark_files_at_shutdown` function that would seem also to be redundant (removed).

Wasn't sure which `fil_flush_file_spaces` calls you were referring to so left these.
Haven't implemented a global time base in `buf_flush_remove`.